### DISCUSSION TITLE
Get list of bond indices from substructure search?

### DISCUSSION THREAD

**Discussion by apahl** [rdkit/rdkit#6650] [Q&A] [answered]
Hi, I would like to apply some multi-colored substructure highlighting to some molecules, based on results from substructure matches.

IIUC, in order to do that, I need to pass two color dictionaries to either `DrawMolecule` or `DrawMoleculeWithHighlights`. One for the atom indices and one for the bond indices. The atom indices I can get from `GetSubstructMatches`.  
Is there a straightforward way to get the corresponding bond indices for a given substructure match?

Thanks a lot.

KR Axel

**Comment by DavidACosgrove** [comment]
Hi Axel,
I use this:
```
def matchedBonds(qmol, mol):
    matchAts = mol.GetSubstructMatch(qmol)
    matchBnds = []
    for ma1 in matchAts:
        for ma2 in matchAts:
            if ma1 > ma2:
                bond = mol.GetBondBetweenAtoms(ma1, ma2)
                if bond is not None:
                    matchBnds.append(bond.GetIdx())
    return matchBnds
```
Although it's not exactly correct because if there are 2 disconnected fragments in the query, separated by a bond that isn't, the bond will still be drawn.  But that's an edge case I live with.  If you want to be completely correct, you would have to do something similar on the matching atoms in the qmol and map back onto the corresponding atoms in the mol.
Dave

**Comment by bp-kelley** [comment, marked as answer]
As @DavidACosgrove said, a better way of doing this is by simply using the query bonds and mapping to the matched molecule, remember that the match list is in order of the query atoms.

```
from rdkit import Chem

def get_bonds_from_match(query_mol, mol, atom_match):
    """
    Args:
      query_mol: query molecule used to match
      mol: molecule matched
      atom_match: result of GetSubstructMatch (i.e. matched atom idx list)
    Returns:
      list of matched bond indices, or None
    """
    if not match: return None
    bonds = []
    for bond in query_mol.GetBonds():
        idx1, idx2 = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
        bonds.append(mol.GetBondBetweenAtoms(atom_match[idx1], atom_match[idx2]).GetIdx())
    return bonds
```

Example

```
>>> mol = Chem.MolFromSmiles("OOC1NC1")
>>> query = Chem.MolFromSmarts("[#6][#7][#6]")

>>> match = mol.GetSubstructMatch(query)
>>> print(get_bonds_from_match(query, mol, match))
[2,3]
```

**Comment by apahl** [comment]
Dear David and Brian, thanks a lot for the very helpful answers! This was exactly what I was looking for.
